### PR TITLE
[ZEPPELIN-3377] Passing Z variables to JDBC interpreter

### PR DESCRIPTION
**Pull request by sanjaydasgupta**
### What is this PR for?
This PR enables the interpolation of ZeppelinContext objects into the paragraph text of JDBC cells. It also introduces a new interpreter-level configuration parameter named _zeppelin.jdbc.interpolation_. This new parameter is _false_ by default, and must be set to _true_ to enable object interpolation. The default value of _false_ guarantees backward compatibility for users who are not aware of the new feature.

The implementation takes the same approach that was followed in [PR-2898](https://github.com/apache/zeppelin/pull/2898).

I have also taken the liberty to correct a preexisting error in the description of the use of Dynamic Forms in the associated documentation (_jdbc.md_).

### What type of PR is it?
[Feature]

### Todos
* [ ] - Task

### What is the Jira issue?
https://issues.apache.org/jira/browse/ZEPPELIN-3377

### How should this be tested?
CI Pass. 

The code in this PR merely causes the JDBC interpreter to "opt-in" to the implementation already existing in the `Interpreter` base class - described in [PR-2898](https://github.com/apache/zeppelin/pull/2898). The unit tests necessary are already present in PR-2898

### Questions:
* Does the licenses files need update? No
* Is there breaking changes for older versions? No
* Does this needs documentation? Yes, documentation has been added to the file _jdbc.md_. I have also taken the liberty to correct a preexisting error in the description of the use of Dynamic Forms in the associated documentation.

### CODE REVIEW

File: docs/interpreter/jdbc.md [hunk 1]
```diff
@@ -240,7 +244,7 @@ You can leverage [Zeppelin Dynamic Form](../usage/dynamic_form/intro.html) insid
 %jdbc_interpreter_name
 SELECT name, country, performer
 FROM demo.performers
-WHERE name='{{"{{performer=Sheryl Crow|Doof|Fanfarlo|Los Paranoia"}}}}'
+WHERE name='${performer=Sheryl Crow|Doof|Fanfarlo|Los Paranoia}'
```
Comment (felixcheung) [inline docs/interpreter/jdbc.md]: I think this is different, right, why this change?

Comment (sanjaydasgupta) [reply docs/interpreter/jdbc.md]: Yes, this is a different thing. I have mentioned under **What is this PR for?**:

> I have also taken the liberty to correct a preexisting error in the description of the use of
> Dynamic Forms in the associated documentation (jdbc.md).

Since I was editing _jdbc.md_, I thought correcting the error would be useful for readers. The changed text works as intended under the current implementation, and is also according to the documentation for Dynamic Forms.

Please let me know if corrections should be handled in a different way. 

Thanks for noticing this update.

Comment (felixcheung) [reply docs/interpreter/jdbc.md]: ok thanks

File: docs/interpreter/jdbc.md [hunk 2]
```diff
@@ -723,6 +727,26 @@ Before Adding one of the below dependencies, check the Phoenix version first.
 </table>
 
 [Maven Repository: org.apache.tajo:tajo-jdbc](https://mvnrepository.com/artifact/org.apache.tajo/tajo-jdbc)
+## Object Interpolation
```
Comment (felixcheung) [inline docs/interpreter/jdbc.md]: I think you need an empty line before `##` or it will not be treated as header properly

File: docs/interpreter/jdbc.md [hunk 3]
```diff
@@ -723,6 +727,26 @@ Before Adding one of the below dependencies, check the Phoenix version first.
...
+The following example shows one use of this facility:
+
+####In Scala cell:
+```
+z.put("country_code", "KR")
+    // ...
+```
+
+####In later JDBC cell:
+```
```
Comment (felixcheung) [inline docs/interpreter/jdbc.md]: add `sql` - see [L230](https://github.com/apache/zeppelin/pull/2903/files#diff-12ee03bf05e397a25720e23c90890661R230)

File: docs/interpreter/jdbc.md [hunk 4]
```diff
@@ -723,6 +727,26 @@ Before Adding one of the below dependencies, check the Phoenix version first.
...
+%jdbc_interpreter_name select * from patents_list where
```
Comment (felixcheung) [inline docs/interpreter/jdbc.md]: newline after `%jdbc_interpreter_name` for clarity & consistency with other example?

File: docs/interpreter/jdbc.md [hunk 5]
```diff
@@ -723,6 +727,26 @@ Before Adding one of the below dependencies, check the Phoenix version first.
...
+####In Scala cell:
+```
+z.put("country_code", "KR")
+    // ...
+```
+
+####In later JDBC cell:
+```
+%jdbc_interpreter_name select * from patents_list where 
+    priority_country = '{country_code}' and filing_date like '2015-__-__'
```
Comment (felixcheung) [inline docs/interpreter/jdbc.md]: is `_` in `2015-__-__` intentional?
perhaps better if the example is `2015-%`? or just pick `2015-01-01`

File: docs/interpreter/jdbc.md (hunk 2 above)
Comment (sanjaydasgupta) [reply docs/interpreter/jdbc.md]: Added preceding blank line.

File: docs/interpreter/jdbc.md (hunk 3 above)
Comment (sanjaydasgupta) [reply docs/interpreter/jdbc.md]: Added the 'sql'

File: docs/interpreter/jdbc.md (hunk 4 above)
Comment (sanjaydasgupta) [reply docs/interpreter/jdbc.md]: Done.

File: docs/interpreter/jdbc.md (hunk 5 above)
Comment (sanjaydasgupta) [reply docs/interpreter/jdbc.md]: I've preferred '2015-%' as using '2015-01-01' would retrieve a too small sample - and not be a realistic example.

**Comment by sanjaydasgupta** [comment]
@felixcheung I've fixed the 4 issues in the documentation file jdbc.md, thanks for the very careful review.

I have also added a unit test [JDBCInterpreterInterpolationTest.java](https://github.com/apache/zeppelin/pull/2903/files#diff-f8ebbd56b41630cac1cec76efe751d9d) which tests the following:

1) enabling and disabling the feature by setting "zeppelin.jdbc.interpolation", including the default value of false.
2) normal substitution of {...} patterns
3) escaping the pattern functionality by using {{...}}

Please also note that all the unit tests defined for this feature in the `Interpreter` base class do a thorough check of all possible uses and misuses of { and }. The above additional unit test is really more like an integration test to verify that the JDBC interpreter is inheriting and leveraging the base class's functionality correctly.

Thanks for your very helpful review.

**Comment by sanjaydasgupta** [comment]
Apologies for missing the check style errors @felixcheung 

The style issues in _JDBCInterpreterInterpolationTest.java_ have been removed, and all the tests now pass. The other failures in the build are unrelated to this feature.

Thanks  for your review.

**Comment by felixcheung** [comment]
still have errors on this link https://travis-ci.org/sanjaydasgupta/zeppelin/builds/363808230?

**Comment by sanjaydasgupta** [comment]
Oops! I must have messed up somewhere!

I've corrected the style issues now @felixcheung, and the one remaining failure appears to be a build script problem unrelated to my code. 

Thanks!

**Comment by sanjaydasgupta** [comment]
ping @felixcheung

**Comment by sanjaydasgupta** [comment]
Closing and reopening to trigger tests.

**Comment by sanjaydasgupta** [comment]
Hi @felixcheung please let me know if any additional changes or adjustments are needed.

The two failing tests are unrelated to the code in in this PR

Thanks.

**Comment by felixcheung** [comment]
looks good!
merging if no more comment

**Comment by sanjaydasgupta** [comment]
Hi @felixcheung is it possible to release this into 0.8.0?

Thanks

**Comment by felixcheung** [comment]
I think so, @zjffdu what do you think about this in 0.8?

**Comment by felixcheung** [comment]
merged this to master & 0.8
